src/pip/_internal/utils/virtualenv.py

```diff
@@ -1,34 +1,107 @@
-import os.path
+from __future__ import absolute_import
+
+import logging
+import os
 import site
 import sys
 
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import List, Optional
+
+logger = logging.getLogger(__name__)
+
+
+def _running_under_venv():
+    # type: () -> bool
+    """Checks if sys.base_prefix and sys.prefix match.
+
+    This handles PEP 405 compliant virtual environments.
+    """
+    return sys.prefix != getattr(sys, "base_prefix", sys.prefix)
+
+
+def _running_under_regular_virtualenv():
+    # type: () -> bool
+    """Checks if sys.real_prefix is set.
+
+    This handles virtual environments created with pypa's virtualenv.
+    """
+    # pypa/virtualenv case
+    return hasattr(sys, 'real_prefix')
+
 
 def running_under_virtualenv():
     # type: () -> bool
+    """Return True if we're running inside a virtualenv, False otherwise.
     """
-    Return True if we're running inside a virtualenv, False otherwise.
+    return _running_under_venv() or _running_under_regular_virtualenv()
+
 
+def _get_pyvenv_cfg_lines():
+    # type: () -> Optional[List[str]]
+    """Reads {sys.prefix}/pyvenv.cfg and returns its contents as list of lines
+
+    Returns None, if it could not read/access the file.
     """
-    if hasattr(sys, 'real_prefix'):
-        # pypa/virtualenv case
-        return True
-    elif sys.prefix != getattr(sys, "base_prefix", sys.prefix):
-        # PEP 405 venv
+    pyvenv_cfg_file = os.path.join(sys.prefix, 'pyvenv.cfg')
+    try:
+        with open(pyvenv_cfg_file) as f:
+            return f.read().splitlines()  # avoids trailing newlines
+    except IOError:
+        return None
+
+
+def _no_global_under_venv():
+    # type: () -> bool
+    """Check `{sys.prefix}/pyvenv.cfg` for system site-packages inclusion
+
+    PEP 405 specifies that when system site-packages are not supposed to be
+    visible from a virtual environment, `pyvenv.cfg` must contain the following
+    line:
+
+        include-system-site-packages = false
+
+    Additionally, log a warning if accessing the file fails.
+    """
+    cfg_lines = _get_pyvenv_cfg_lines()
+    if cfg_lines is None:
+        # We're not in a "sane" venv, so assume there is no system
+        # site-packages access (since that's PEP 405's default state).
+        logger.warning(
+            "Could not access 'pyvenv.cfg' despite a virtual environment "
+            "being active. Assuming global site-packages is not accessible "
+            "in this environment."
+        )
         return True
 
-    return False
+    return "include-system-site-packages = false" in cfg_lines
```

Comment:
I'd prefer a more robust parsing (either with `configparser`, or `re`, or any code that would accept `include-system-site-packages=false` and its whitespace variants...)

Reply:
Well, AFAICT, that sort of whitespace is not allowed in PEP 405 config files.

> this file is scanned for lines of the form `key = value`

Reply:
Whoops! Here's the reference I forgot to paste in the previous comment -- That sentence is from the second paragraph in the Specification section of PEP 405: https://www.python.org/dev/peps/pep-0405/#specification

Reply:
My reading of the PEP doesn't lead me to think the whitespace is mandatory. And a quick experiment confirms that if I manually edit `pyvenv.cfg` to say `include-system-site-packages=true` (no whitespace) it does still make the system site packages visible.

So +1 on making the test ignore whitespace.

Reply:
Cool cool. No worries from my end on this.

I just didn't do it coz it didn't feel important.